rust/yarp-sys/build.rs

```diff
@@ -0,0 +1,180 @@
+use std::{
+    io::{self, Write},
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+fn main() -> Result<(), io::Error> {
+    rake_compile()?;
+
+    let ruby_build_path = ruby_build_path();
+    let ruby_include_path = ruby_include_path();
+
+    // Tell cargo/rustc that we want to link against `librubyparser.a`.
+    println!("cargo:rustc-link-lib=static=rubyparser");
+
+    // Add `[root]/build/` to the search paths, so it can find `librubyparser.a`.
+    println!(
+        "cargo:rustc-link-search=native={}",
+        ruby_build_path.display()
+    );
+
+    // This is where the magic happens.
+    let bindings = generate_bindings(&ruby_include_path);
+
+    // Write the bindings to file.
+    write_bindngs(bindings);
+
+    Ok(())
+}
+
+/// Convenience step for running `rake compile` before doing our stuff.
+///
+fn rake_compile() -> Result<(), io::Error> {
+    let output = Command::new("rake").arg("compile").output()?;
+
+    io::stdout().write_all(&output.stdout)?;
+    io::stderr().write_all(&output.stderr)?;
+
+    println!("`rake compile` exit status: {}", output.status);
+
+    Ok(())
+}
+
+/// Gets the path to project files (`librubyparser*`) at `[root]/build/`.
+///
+fn ruby_build_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../build/")
+        .canonicalize()
+        .unwrap()
+}
+
+/// Gets the path to the header files that `bindgen` needs for doing code generation.
+///
+fn ruby_include_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../include/")
+        .canonicalize()
+        .unwrap()
+}
+
+fn cargo_manifest_path() -> &'static Path {
+    Path::new(env!("CARGO_MANIFEST_DIR"))
+}
+
+/// Uses `bindgen` to generate bindings to the C API. Update this to allow new types/functions/etc
+/// to be generated (it's allowlisted to only expose functions that'd make sense for public
+/// consumption).
+///
+/// This method only generates code in memory here--it doesn't write it to file.
+///
+fn generate_bindings(ruby_include_path: &Path) -> bindgen::Bindings {
+    bindgen::Builder::default()
```

Comment:
My approach here was to only expose things in the public C interface. If you don't do any of the `allowlist_*`/`blocklist_*` stuff, `bindgen` will do bindings for everything it finds in the headers you give it.